packages/tpetra/core/test/CrsGraph/CrsGraph_PackUnpack.cpp

```diff
@@ -0,0 +1,419 @@
+//                 Copyright (2008) Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+*/
+
+#include "Tpetra_TestingUtilities.hpp"
+#include "TpetraCore_ETIHelperMacros.h"
+#include "Tpetra_CrsGraph.hpp"
+#include "Tpetra_DefaultPlatform.hpp"
+#include "Tpetra_Distributor.hpp"
+#include "Tpetra_Map.hpp"
+#include "Tpetra_Details_gathervPrint.hpp"
+#include "Tpetra_Details_packCrsGraph.hpp"
+#include "Tpetra_Details_unpackCrsGraphAndCombine.hpp"
+#include "Teuchos_CommHelpers.hpp"
+#include "Kokkos_ArithTraits.hpp"
+#include <random>
+#include <set>
+
+namespace { // anonymous
+
+#define NUM_ROW_PER_PROC 4
+#define NUM_NZ_COLS 100
+
+using Tpetra::TestingUtilities::getDefaultComm;
+using Teuchos::rcp;
+using Teuchos::RCP;
+using Teuchos::Array;
+using Teuchos::ArrayView;
+using Teuchos::Comm;
+using Teuchos::outArg;
+using Tpetra::Details::gathervPrint;
+using Tpetra::Details::packCrsGraph;
+using Tpetra::Details::unpackCrsGraphAndCombine;
+using std::endl;
+
+template<class T>
+bool
+essentially_equal(T a, T b) {
+  typedef Kokkos::ArithTraits<T> KAT;
+  const auto eps = KAT::eps();
+  return KAT::abs(a - b) <= ( (KAT::abs(a) > KAT::abs(b) ? KAT::abs(b) : KAT::abs(a)) * eps);
```

Comment:
Fun fact: this should also return false for a or b NaN, which is what you want :-) .